Add more tests for userfaultfd

ricardobranco777 opened the pull request:
Add more tests for userfaultfd:
- UFFDIO_MOVE
- UFFDIO_WRITEPROTECT
- UFFDIO_ZEROPAGE
- UFFD_USER_MODE_ONLY
- Use /dev/userfaultfd instead of system call

metan-ucw commented on testcases/kernel/syscalls/userfaultfd/userfaultfd02.c:
```diff
@@ -0,0 +1,119 @@
+	SAFE_IOCTL(uffd, UFFDIO_REGISTER, &uffdio_register);
+
+	SAFE_PTHREAD_CREATE(&thr, NULL,
+			(void * (*)(void *)) handle_thread, NULL);
```
This can be just (void*) cast, but it would be better if the handle_thread function was defined as void * and returned NULL at the end.

metan-ucw commented on testcases/kernel/syscalls/userfaultfd/userfaultfd05.c:
```diff
@@ -0,0 +1,116 @@
+	SAFE_PTHREAD_CREATE(&thr, NULL,
+			(void * (*)(void *)) handle_thread, NULL);
+
+	(void)page[0xf];
```
I do not think that we need this here, we are checking the page in the loop below, which will trigger read access anyways.

ricardobranco777 commented on testcases/kernel/syscalls/userfaultfd/userfaultfd03.c:
```diff
@@ -0,0 +1,114 @@
+static struct tst_test test = {
+	.test_all = run,
+	.min_kver = "5.11",
+};
```
The `.min_kver` is different though.

ricardobranco777 commented:
> Looks good, but there is too much duplicated code (should be mostly put into a common header) and some minor errors as well.

Thanks for reviewing.  I addressed most of your suggestions.  Left some questions regarding further refactoring.

pevik commented on testcases/kernel/syscalls/userfaultfd/userfaultfd02.c:
```diff
@@ -0,0 +1,105 @@
+ * from a different thread using UFFDIO_MOVE
+ */
+
+#include "config.h"
```
nit (for all tests): we don't need `config.h` (this is needed only if `#ifdef` or `#if` preprocessor extensions are used, which use values from `include/config.h`.

pevik commented on testcases/kernel/syscalls/userfaultfd/tst_userfaultfd.h:
```diff
@@ -0,0 +1,23 @@
+				"Could not create userfault file descriptor");
+	}
+
+	return TST_RET;
```
`sys_userfaultfd()` is not needed any more since 98d1af9f508f93a3cfb39b9c32c9aaa91bc8e196, which added `SAFE_USERFAULTFD()`. Please rebase and use it.

ricardobranco777 commented:
> Thanks! Just join userfaultfd0[13].c is missing.

I've just merged the `UFFD_USER_MODE_ONLY` test into userfaultfd01.c. 

We can enhance `userfaultfd0{2,4,5}.c` to have a `UFFD_USER_MODE_ONLY` case like `userfaultfd01.c`. WDYT?

pevik commented on testcases/kernel/syscalls/userfaultfd/userfaultfd02.c:
```diff
@@ -0,0 +1,100 @@
+	SAFE_READ(1, uffd, &msg, sizeof(msg));
+
+	if (msg.event != UFFD_EVENT_PAGEFAULT)
+		tst_brk(TBROK | TERRNO, "Received unexpected UFFD_EVENT");
```
nit: Maybe print wrong event?
```c
tst_brk(TBROK | TERRNO, "Received unexpected UFFD_EVENT: %d", msg.event);

pevik replied:
My suggestion above would also print `msg.event` (the invalid number). `errno` is useful, but invalid number as well (it helps to identify where the problem is).

ricardobranco777 replied:
Fixed!

pevik commented:
I guess all has been fixed, but give @metan-ucw chance for final ack of the changes. I plan to merge tomorrow afternoon (or after his ack).
